### PR TITLE
Fix redirecting to `/search` 

### LINKED ISSUE #78: `/` route is changed to `/search`
The spec from #69 was

> Proxy to `/` to `/search` but still appears as `/`. 

But it now appears as /search

### PR DESCRIPTION
**Pull request by AbdulrhmnGhanem**
Fixes #78

### CODE REVIEW

**Comment by AbdulrhmnGhanem** [comment]
@kasbah, could you review the functionality of this? Just the functionality, I will improve the implementation.

**Comment by kasbah** [comment]
This has some odd behaviour. Once you submit a search it seems to go back to updating while you type.

**Comment by AbdulrhmnGhanem** [comment]
>  Once you submit a search it seems to go back to updating while you type.

It is either
1. `revalidateOnFocus`. Disabled it in 020613a.
2. clearing the search form; it resets the projects to all projects, so it hits the endpoint. Made it rely on SWR instead of next in 9f047ef.

## No requests while typing
image

**Comment by kasbah** [comment]
I don't understand all the details here but it still has the odd behaviour I described (currently [deployed to kaspar-dev](https://kaspar-dev.staging.kitspace.dev)). Please make this work like a normal sensible search.

File: frontend/cypress/integration/search.spec.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  it('should use `q` from query parameters', () => {
+    const username = faker.unique(faker.name.firstName)
+    const email = faker.unique(faker.internet.email)
+    const password = '123456'
+
+    const repoName = 'HACK'
+    const syncedRepoUrl = 'https://github.com/kitspace-forks/HACK'
+
+    cy.intercept('http://gitea.kitspace.test:3000/user/kitspace/**').as('sign_in')
```
Comment (kasbah) [inline frontend/cypress/integration/search.spec.js:81]: Please don't add new e2e test code that relies on backend requests.

**Comment by kasbah** [comment]
(This is also occasionally showing a 404 page on the index page on kaspar-dev).

File: frontend/cypress/integration/search.spec.js (hunk 1 above)
Comment (AbdulrhmnGhanem) [reply frontend/cypress/integration/search.spec.js:81]: This is one of the reasons I told you choosing a base is getting tricky.
I removed all interceptions in #179. You can see search.spec [here](https://github.com/kitspace/kitspace-v2/blob/c04f0ab75eb768ee8f01f7d998555a4fc01db745/frontend/cypress/integration/search.spec.js).

**Comment by AbdulrhmnGhanem** [comment]
> (This is also occasionally showing a 404 page on the index page on kaspar-dev).

image

I am not sure what is going on here. Could you look at the logs, please?

**Comment by kasbah** [comment]
I don't have time, sorry. Can you deploy yourself and debug please?

**Comment by AbdulrhmnGhanem** [comment]
I did so, it isn't reproducible.

**Comment by kasbah** [comment]
Yep, haha. I've also given you access to `ssh ubuntu@kaspar-dev.staging.kitspace.dev` so you can take a look at the errors that occur there .

**Comment by kasbah** [comment]
I've deployed this to https://kaspar-dev.staging.kitspace.dev/ and the search doesn't seem to work at all.

**Comment by kasbah** [comment]
Oh, it's mainly just really slow. Seemingly because of processor response times :thinking:

File: frontend/src/contexts/SearchContext.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const updateQuery = q => {
+    setQuery(q)
+  }
+
+  useEffect(() => {
+    if (q == null) setQuery('')
+  }, [q])
+
+  return (
+    <SearchContext.Provider value={{ query, updateQuery }}>
```
Comment (kasbah) [inline frontend/src/contexts/SearchContext.js:24]:
```suggestion
<SearchContext.Provider value={{ query, updateQuery: setQuery }}>
```

File: frontend/src/contexts/SearchContext.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+const SearchProvider = ({ children, initialQuery }) => {
+  const [query, setQuery] = useState(initialQuery ?? '')
+
+  const {
+    query: { q },
+  } = useRouter()
+  const updateQuery = q => {
+    setQuery(q)
+  }
```
Comment (kasbah) [inline frontend/src/contexts/SearchContext.js:17]:
```suggestion
```

**Comment by AbdulrhmnGhanem** [comment]
* Remove the submit button; 4d8e798, and bdadb2f.
* Submitting empty form displays all projects; the behaviour was implemented, but it was impossible to submit an empty form because the button was disabled for empty forms.
* flashing data on search submit; a424c34.

File: frontend/cypress/integration/search.spec.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
+import faker from 'faker'
+
+describe('Homepage search bar', () => {
+  it('should redirect to /search on clicking `Search` button', () => {
```
Comment (kasbah) [inline frontend/cypress/integration/search.spec.js:4]:
```suggest
  it('should redirect to /search when search is submitted', () => {
```

**Pull request merged**